### PR TITLE
o2-eve-display: added TRD, TOF, MID to the workflow

### PR DESCRIPTION
**Pull request by jmyrcha**
added information fields to *.json file to recognise workflow parameters used to produce this json

### CODE REVIEW

**Comment by jmyrcha** [comment]
fail of build/O2/o2-cs8 is not related to branch code

File: EventVisualisation/DataConverter/include/EventVisualisationDataConverter/VisualisationEvent.h [hunk 1]
```diff
@@ -106,19 +106,14 @@ class VisualisationEvent
...
-
-  // Returns number of clusters
-  size_t getClusterCount() const
-  {
-    return mClusters.size();
-  }
+  const VisualisationCluster& getCluster(int i) const { return mClusters[i]; };
+  size_t getClusterCount() const { return mClusters.size(); } // Returns number of clusters
+  void setWorkflowVersion(float workflowVersion) { this->mWorkflowVersion = workflowVersion; }
+  void setWorkflowParameters(std::string workflowParameters) { this->mWorkflowParameters = workflowParameters; }
```
Comment (shahor02) [inline EventVisualisation/DataConverter/include/EventVisualisationDataConverter/VisualisationEvent.h]: please change this to `const std::string& workflowParameters`

File: EventVisualisation/Workflow/include/EveWorkflow/EveWorkflowHelper.h [hunk 2]
```diff
@@ -60,33 +63,40 @@ class EveWorkflowHelper
...
  public:
   EveWorkflowHelper();
   static std::vector<PNT> getTrackPoints(const o2::track::TrackPar& trc, float minR, float maxR, float maxStep, float minZ = 25000, float maxZ = 25000);
   void selectTracks(const CalibObjectsConst* calib, GID::mask_t maskCl,
                     GID::mask_t maskTrk, GID::mask_t maskMatch);
-  template <typename Functor>
-  void addTrackToEvent(Functor source, GID gid, float trackTime, float z = 0.); // store track in mEvent
-  void draw(std::string jsonPath, int numberOfFiles, int numberOfTracks = -1);
+  void addTrackToEvent(const o2::track::TrackParCov& tr, GID gid, float trackTime, float dz);
+  void draw(std::string jsonPath, int numberOfFiles, int numberOfTracks, o2::dataformats::GlobalTrackID::mask_t trkMask, o2::dataformats::GlobalTrackID::mask_t clMask, float mWorkflowVersion);
```
Comment (shahor02) [inline EventVisualisation/Workflow/include/EveWorkflow/EveWorkflowHelper.h]: here too

File: EventVisualisation/Workflow/src/EveWorkflowHelper.cxx [hunk 3]
```diff
@@ -44,7 +43,9 @@ void EveWorkflowHelper::selectTracks(const CalibObjectsConst* calib,
   this->mRecoCont.createTracksVariadic(creator);
 }
 
-void EveWorkflowHelper::draw(std::string jsonPath, int numberOfFiles, int numberOfTracks)
+void EveWorkflowHelper::draw(std::string jsonPath, int numberOfFiles, int numberOfTracks,
```
Comment (shahor02) [inline EventVisualisation/Workflow/src/EveWorkflowHelper.cxx]: here too

File: EventVisualisation/Workflow/src/EveWorkflowHelper.cxx [hunk 4]
```diff
@@ -171,28 +193,140 @@ void EveWorkflowHelper::prepareMFTClusters(const o2::itsmft::TopologyDictionary&
...
+  const auto detRefs = mRecoCont.getSingleDetectorRefs(gid);
+  if (detRefs[o2::dataformats::GlobalTrackID::Source::ITS].isIndexSet()) {
+    drawITSClusters(detRefs[o2::dataformats::GlobalTrackID::Source::ITS], trackTime);
+  }
+  if (detRefs[o2::dataformats::GlobalTrackID::Source::ITSAB].isIndexSet()) {
+    drawITSClusters(detRefs[o2::dataformats::GlobalTrackID::Source::ITSAB], trackTime);
+  }
+  if (detRefs[o2::dataformats::GlobalTrackID::Source::TPC].isIndexSet()) {
+    drawTPCClusters(detRefs[o2::dataformats::GlobalTrackID::Source::TPC], trackTime);
+  }
```
Comment (shahor02) [inline EventVisualisation/Workflow/src/EveWorkflowHelper.cxx]: as I wrote you, there is no point in testing `isIndexSet()` here, TPCITS tracks always has TPC part and always has either ITS or ITSAB. And the `drawITSClusters` knows how to handle both. You even don't need to do getRefs, which is a relatively expensive function. The previous code was correct, just do
```
drawITSClusters(track.getRefITS(), trackTime);
drawTPCClusters(track.getRefTPC(), trackTime);
```

File: EventVisualisation/Workflow/src/EveWorkflowHelper.cxx [hunk 5]
```diff
@@ -171,28 +193,140 @@ void EveWorkflowHelper::prepareMFTClusters(const o2::itsmft::TopologyDictionary&
...
+
+void EveWorkflowHelper::drawTPCTRD(GID gid, float trackTime)
+{
+  //LOG(INFO) << "EveWorkflowHelper::drawTPCTRD " << gid;
+
+  const auto detRefs = mRecoCont.getSingleDetectorRefs(gid);
+  if (detRefs[o2::dataformats::GlobalTrackID::Source::TPC].isIndexSet()) {
+    const auto& tpcTrack = mRecoCont.getTPCTrack(detRefs[o2::dataformats::GlobalTrackID::Source::TPC]);
+    addTrackToEvent(tpcTrack, gid, trackTime, 0);
+    drawTPCClusters(detRefs[o2::dataformats::GlobalTrackID::Source::TPC], trackTime);
```
Comment (shahor02) [inline EventVisualisation/Workflow/src/EveWorkflowHelper.cxx]: here too, no need for `getSingleDetectorRefs` and `isIndexSet`, just do
```
const auto& tpcTrdTrack = mRecoCont.getTPCTRDTrack<o2::trd::TrackTRD>(gid);
drawTPCClusters(tpcTrdTrack.getRefGlobalTrackId(), trackTime);
```

File: EventVisualisation/Workflow/src/EveWorkflowHelper.cxx [hunk 6]
```diff
@@ -171,28 +193,140 @@ void EveWorkflowHelper::prepareMFTClusters(const o2::itsmft::TopologyDictionary&
...
+  const auto detRefs = mRecoCont.getSingleDetectorRefs(gid);
+  if (detRefs[o2::dataformats::GlobalTrackID::Source::ITS].isIndexSet()) {
+    drawITSClusters(detRefs[o2::dataformats::GlobalTrackID::Source::ITS], trackTime);
+  }
+  if (detRefs[o2::dataformats::GlobalTrackID::Source::ITSAB].isIndexSet()) {
+    drawITSClusters(detRefs[o2::dataformats::GlobalTrackID::Source::ITSAB], trackTime);
+  }
+  if (detRefs[o2::dataformats::GlobalTrackID::Source::TPC].isIndexSet()) {
+    drawTPCClusters(detRefs[o2::dataformats::GlobalTrackID::Source::TPC], trackTime);
+  }
```
Comment (shahor02) [inline EventVisualisation/Workflow/src/EveWorkflowHelper.cxx]: same here, bit you can have it even shorter:
```
drawITSTPC(itsTpcTrdTrack.getRefGlobalTrackId(), trackTime);

```

File: EventVisualisation/Workflow/src/EveWorkflowHelper.cxx [hunk 7]
```diff
@@ -171,28 +193,140 @@ void EveWorkflowHelper::prepareMFTClusters(const o2::itsmft::TopologyDictionary&
...
+  drawTRDClusters(itsTpcTrdTofTrack, trackTime);
```
Comment (shahor02) [inline EventVisualisation/Workflow/src/EveWorkflowHelper.cxx]: this is wrong: `gid` here points on the entry in the TOF matches, not in the TPCTRDTracks, should be 
```
const auto& match = mRecoCont.getITSTPCTRDTOFMatches()[gid.getIndex()];
auto gidITSTPCTRD = match.getTrackRef();
```
then either draw it using `drawITSTPCTRD` or implement similar code here.

File: EventVisualisation/Workflow/src/EveWorkflowHelper.cxx [hunk 8]
```diff
@@ -171,28 +193,140 @@ void EveWorkflowHelper::prepareMFTClusters(const o2::itsmft::TopologyDictionary&
...
+
+void EveWorkflowHelper::drawTPCTRDTOF(GID gid, float trackTime)
+{
+  // LOG(INFO) << "EveWorkflowHelper::drawTPCTRDTOF " << gid;
+  const auto& tpcTrdTofTrack = mRecoCont.getTPCTRDTrack<o2::trd::TrackTRD>(gid);
+  addTrackToEvent(tpcTrdTofTrack, gid, trackTime, 0);
+  const auto detRefs = mRecoCont.getSingleDetectorRefs(gid);
+  if (detRefs[o2::dataformats::GlobalTrackID::Source::TPC].isIndexSet()) {
+    drawTPCClusters(detRefs[o2::dataformats::GlobalTrackID::Source::TPC], trackTime);
+  }
```
Comment (shahor02) [inline EventVisualisation/Workflow/src/EveWorkflowHelper.cxx]: this is also wrong: 
```
const auto& match = mRecoCont.getITPCTRDTOFMatches()[gid.getIndex()];
auto gidITPCTRD = match.getTrackRef();
```
then either draw it using drawTPCTRD or implement similar code here.

File: EventVisualisation/Workflow/src/EveWorkflowHelper.cxx [hunk 9]
```diff
@@ -171,28 +193,140 @@ void EveWorkflowHelper::prepareMFTClusters(const o2::itsmft::TopologyDictionary&
...
+
+void EveWorkflowHelper::drawTPCTOF(GID gid, float trackTime)
+{
+  //  LOG(INFO) << "EveWorkflowHelper::drawTPCTRDTOF " << gid;
+  const auto detRefs = mRecoCont.getSingleDetectorRefs(gid);
+  if (detRefs[o2::dataformats::GlobalTrackID::Source::TPC].isIndexSet()) {
+    const auto& tpcTrack = mRecoCont.getTPCTrack(detRefs[o2::dataformats::GlobalTrackID::Source::TPC]);
+    addTrackToEvent(tpcTrack, gid, trackTime, 0);
+    drawTPCClusters(detRefs[o2::dataformats::GlobalTrackID::Source::TPC], trackTime);
+  }
```
Comment (shahor02) [inline EventVisualisation/Workflow/src/EveWorkflowHelper.cxx]: also wrong, as I wrote you, TPCTOF are refitted tracks with their own kinematics.
```
const auto& trTPCTOF = mRecoCont.getTPCTOFTrack(gid);
const auto& match = mRecoCont.getTPCTOFMatch(gid.getIndex());
addTrackToEvent(trTPCTOF, gid, trackTime, 0);
drawTPCClusters(match.getTrackRef(), trackTime);
```

**Review by shahor02 — CHANGES_REQUESTED** [review]

**Pull request merged**